surveySimPP/modules/PPResolveMagnitudeInFilter.py

```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/python
+
+import pandas as pd
+import numpy as np
+
+# Author: Grigori Fedorets, Meg Schwamb
+
+
+def PPResolveMagnitudeInFilter(padain,mainfilter,othercolours,resfilters):
+    """
+    PPResolveMagnitudeInFilter.py
+    
+    Description: This tasks selects a colour offset relevant to each filter at each given pointing
+    and calculates the colour in each given filter. The apparent magnitude has already been
+    calculated in the main filter.
+    
+    
+    Mandatory input: string, padain, name of input pandas dataframe
+                     string, mainfilter, name of the main filter in which the apparent magnitude has been calculated
+                     array of strings, othercolours, names of colour offsets (e.g. r-i)
+                     array of strings, resfilters, names of resulting colours, main filter is the first one, followed 
+                     in order by resolved colours, such as, e.g. 'r'+'g-r'='g'. They should be given in the following order: 
+                     main filter, resolved filters in the same order as respective other colours.
+    
+    Output: updated padain
+    
+    Usage: padaout=PPResolveMagnitudeInFilter(padain,mainfilter,othercolours,resfilters)
+    
+    """
+    
+    apparent_mag=np.zeros(len(padain), dtype=float)
+    
+    i=0
+    while(i<len(resfilters)):
+        inRelevantFilterList=(padain['optFilter']==resfilters[i])
+        inRelevantFilter=padain[inRelevantFilterList]
+        if(len(inRelevantFilter)>0):
+            if(resfilters[i]==mainfilter):
+               apparent_mag[inRelevantFilterList]=0.0
+            else:
+               apparent_mag[inRelevantFilterList]=inRelevantFilter[othercolours[i-1]]
```

Comment:
I don't see how this step works correctly. If i starts out as 0 then in the first time through this you get for the index [i-1]=[0-1]=-1 which means you're polling the last item in the othercolors array when you want 0.  I think this should be i not i-1

Reply:
Can you please check this is working correctly.

Reply:
So, the way this works, is that in the config file we have three different variables:

1. mainfilter, to which everything else is compared
2. othercolours, which are the colour offsets, such as i-r
3. resfilters, which are ther results of colour manipulations. mainfilter is always the first one, with the other resulting filters following in the same order as othercolours. Therefore, len(resfilters) == len(othercolours) +1

The thoughts behind this admitedly somewhat clumsy structures are
1. there needs to be a place in the config file where we can see at a glance which all filters are included
2. there needs to be some way of matching colour offsets and resulting colours.

So for that structure everything should be correct. One thing how this situation can be omitted is pre-calculating all colours in the input parameter file. So instead of g-r, i-r etc we would have precalculated values for g, r, etc.

Reply:
That's dangerous to have resfilters and othercolours have different lengths. It makes it very easy for someone to forget this and do bad things with loops. othercolours[0] =0 for the main filter and the rest of the color terms following in the array would actually significantly simplify the code. You wouldn't need this bit  

`  if(resfilters[i]==mainfilter):
               apparent_mag[inRelevantFilterList]=0.0`

Reply:
If you want to keep resfilter as with no entry for r, then this for loop needs to loop over othercolours and not resfilters. and pull the bit for the main filter out of the loop and do the main filter bit before the loop.  As written you're checking the logic statement each time of whether the main filter is the current filter being looked at in the loop. Whereas if you loop over othercolours and have resfilters[i+1] to access the correct array index then you will speed up the calculation because the logic statement for the main filter isn not being accessed every time.  We need every little speed up we can get

Reply:
How about I combine mainfilter and othercolours into a single array called inputcolours (also in the Config file)? This will force the colours that the user is using in one place, remove the possibility for errors, and get rid of the pesky if-statements?

So, we would have inputcolours=r,g-r,i-r and resfilters r,g,i.

Reply:
Changed per the second suggestion and added to the pull request.